v2.0.0 container unable to start

This is what we get in our first deployment to k8s:

```
[Runner] Cannot find module '/app/node_modules/@verana-labs/verana-types/dist/codec/verana/cs/v1/tx.js.js' imported from /app/dist/src/common/utils/veranaChain.client.js Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/app/node_modules/@verana-labs/verana-types/dist/codec/verana/cs/v1/tx.js.js' imported from /app/dist/src/common/utils/veranaChain.client.js
    at finalizeResolution (node:internal/modules/esm/resolve:275:11)
    at moduleResolve (node:internal/modules/esm/resolve:861:10)
    at defaultResolve (node:internal/modules/esm/resolve:985:11)
    at #cachedDefaultResolve (node:internal/modules/esm/loader:731:20)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:708:38)
    at ModuleLoader.getModuleJobForImport (node:internal/modules/esm/loader:310:38)
    at ModuleJob._link (node:internal/modules/esm/module_job:182:49) {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///app/node_modules/@verana-labs/verana-types/dist/codec/verana/cs/v1/tx.js.js'
}
```

Since we are using Kubernetes, it becomes crucial to regularly check proper Docker image generation and run it locally so we can be sure that the artifact deployed equals the one we are unit testing during development.

--- a/src/scripts/fix-imports.cjs
+++ b/src/scripts/fix-imports.cjs
@@ -57,6 +57,14 @@ function needsJsExtension(importPath) {
   return false;
 }
 
+function normalizeJsExtension(importPath) {
+  return importPath.replace(/(?:\.js)+$/u, ".js");
+}
+
+function withJsExtension(importPath) {
+  return normalizeJsExtension(`${importPath}.js`);
+}
+
 // Fix imports/exports in a single file
 function fixImportsInFile(filePath) {
   try {
@@ -83,23 +91,23 @@ function fixImportsInFile(filePath) {
         if (jsExtType) {
           modified = true;
           if (jsExtType === "index") {
-            return match.replace(importPath, `${importPath}/index.js`);
+            return match.replace(importPath, normalizeJsExtension(`${importPath}/index.js`));
           }
-          return match.replace(importPath, `${importPath}.js`);
+          return match.replace(importPath, withJsExtension(importPath));
         }
 
         // Handle relative imports (./ or ../)
         if (importPath.startsWith("./") || importPath.startsWith("../")) {
           modified = true;
           if (isDirectory(fileDir, importPath)) {
-            return match.replace(importPath, `${importPath}/index.js`);
+            return match.replace(importPath, normalizeJsExtension(`${importPath}/index.js`));
           } else {
-            return match.replace(importPath, `${importPath}.js`);
+            return match.replace(importPath, withJsExtension(importPath));
           }
         }
         if (importPath.startsWith("@verana-labs/verana-types/")) {
           modified = true;
-          return match.replace(importPath, `${importPath}.js`);
+          return match.replace(importPath, withJsExtension(importPath));
         }
         return match; // leave other imports untouched
       }
@@ -119,23 +127,23 @@ function fixImportsInFile(filePath) {
         if (jsExtType) {
           modified = true;
           if (jsExtType === "index") {
-            return match.replace(importPath, `${importPath}/index.js`);
+            return match.replace(importPath, normalizeJsExtension(`${importPath}/index.js`));
           }
-          return match.replace(importPath, `${importPath}.js`);
+          return match.replace(importPath, withJsExtension(importPath));
         }
 
         // Handle relative imports (./ or ../)
         if (importPath.startsWith("./") || importPath.startsWith("../")) {
           modified = true;
           if (isDirectory(fileDir, importPath)) {
-            return match.replace(importPath, `${importPath}/index.js`);
+            return match.replace(importPath, normalizeJsExtension(`${importPath}/index.js`));
           } else {
-            return match.replace(importPath, `${importPath}.js`);
+            return match.replace(importPath, withJsExtension(importPath));
           }
         }
         if (importPath.startsWith("@verana-labs/verana-types/")) {
           modified = true;
-          return match.replace(importPath, `${importPath}.js`);
+          return match.replace(importPath, withJsExtension(importPath));
         }
         return match; // leave other imports untouched
       }

--- a/src/scripts/fix-imports.cjs
+++ b/src/scripts/fix-imports.cjs
@@ -65,6 +65,41 @@ function withJsExtension(importPath) {
   return normalizeJsExtension(`${importPath}.js`);
 }
 
+function normalizeVeranaTypesImport(match, importPath) {
+  if (!importPath.startsWith("@verana-labs/verana-types/")) {
+    return null;
+  }
+
+  const normalizedImportPath = importPath.replace(/(?:\.js)+$/u, "");
+  if (normalizedImportPath !== importPath) {
+    return {
+      match: match.replace(importPath, normalizedImportPath),
+      modified: true,
+    };
+  }
+
+  return { match, modified: false };
+}
+
+function normalizeAlreadyJsImport(match, importPath) {
+  if (importPath.endsWith(".json")) {
+    return { match, modified: false };
+  }
+
+  if (importPath.endsWith(".js")) {
+    const normalizedImportPath = normalizeJsExtension(importPath);
+    if (normalizedImportPath !== importPath) {
+      return {
+        match: match.replace(importPath, normalizedImportPath),
+        modified: true,
+      };
+    }
+    return { match, modified: false };
+  }
+
+  return null;
+}
+
 // Fix imports/exports in a single file
 function fixImportsInFile(filePath) {
   try {
@@ -81,9 +116,16 @@ function fixImportsInFile(filePath) {
     content = content.replace(
       importExportRegex,
       (match, keyword, importPath) => {
-        // Skip if already .js or .json
-        if (importPath.endsWith(".js") || importPath.endsWith(".json")) {
-          return match;
+        const normalizedVeranaTypesImport = normalizeVeranaTypesImport(match, importPath);
+        if (normalizedVeranaTypesImport) {
+          modified = modified || normalizedVeranaTypesImport.modified;
+          return normalizedVeranaTypesImport.match;
+        }
+
+        const normalizedAlreadyJsImport = normalizeAlreadyJsImport(match, importPath);
+        if (normalizedAlreadyJsImport) {
+          modified = modified || normalizedAlreadyJsImport.modified;
+          return normalizedAlreadyJsImport.match;
         }
 
         // Handle known packages
@@ -104,10 +146,6 @@ function fixImportsInFile(filePath) {
           } else {
             return match.replace(importPath, withJsExtension(importPath));
           }
-        }
-        if (importPath.startsWith("@verana-labs/verana-types/")) {
-          modified = true;
-          return match.replace(importPath, withJsExtension(importPath));
         }
         return match; // leave other imports untouched
       }
@@ -117,9 +155,16 @@ function fixImportsInFile(filePath) {
     content = content.replace(
       dynamicImportRegex,
       (match, importPath) => {
-        // Skip if already .js or .json
-        if (importPath.endsWith(".js") || importPath.endsWith(".json")) {
-          return match;
+        const normalizedVeranaTypesImport = normalizeVeranaTypesImport(match, importPath);
+        if (normalizedVeranaTypesImport) {
+          modified = modified || normalizedVeranaTypesImport.modified;
+          return normalizedVeranaTypesImport.match;
+        }
+
+        const normalizedAlreadyJsImport = normalizeAlreadyJsImport(match, importPath);
+        if (normalizedAlreadyJsImport) {
+          modified = modified || normalizedAlreadyJsImport.modified;
+          return normalizedAlreadyJsImport.match;
         }
 
         // Handle known packages
@@ -140,10 +185,6 @@ function fixImportsInFile(filePath) {
           } else {
             return match.replace(importPath, withJsExtension(importPath));
           }
-        }
-        if (importPath.startsWith("@verana-labs/verana-types/")) {
-          modified = true;
-          return match.replace(importPath, withJsExtension(importPath));
         }
         return match; // leave other imports untouched
       }

--- a/src/scripts/global-log-capture.mjs
+++ b/src/scripts/global-log-capture.mjs
@@ -45,7 +45,8 @@ if (!shouldLogToFile) {
 
   const errorDestination = pino.destination({
     dest: errorFilePath,
-    sync: false,
+    // Use sync destination to avoid SonicBoom readiness race at process shutdown.
+    sync: true,
   });
 
   const errorLogger = pino(
@@ -108,20 +109,13 @@ function safeStringify(arg) {
 
   process.on("uncaughtException", (err) => {
     try {
-      const finalLogger = pino.final(errorLogger, (_err, final) => {
-        try {
-          final.fatal(
-            {
-              message: err && err.message,
-              stack: err && err.stack,
-            },
-            "uncaughtException"
-          );
-        } catch (_) {
-          // swallow
-        }
-      });
-      finalLogger(err);
+      errorLogger.fatal(
+        {
+          message: err && err.message,
+          stack: err && err.stack,
+        },
+        "uncaughtException"
+      );
     } catch (_) {}
 
     originalConsole.error(err);
@@ -148,17 +142,12 @@ function safeStringify(arg) {
         );
       }
 
-      const finalLogger = pino.final(errorLogger, (_err, final) => {
-        try {
-          final.error(
-            {
-              reason: reason instanceof Error ? reason.message : safeStringify(reason),
-            },
-            "unhandledRejection - final flush"
-          );
-        } catch (_) {}
-      });
-      finalLogger(reason instanceof Error ? reason : new Error(String(reason)));
+      errorLogger.error(
+        {
+          reason: reason instanceof Error ? reason.message : safeStringify(reason),
+        },
+        "unhandledRejection"
+      );
     } catch (_) {}
 
     originalConsole.error("Unhandled Promise Rejection:", reason);

--- a/src/scripts/global-log-capture.mjs
+++ b/src/scripts/global-log-capture.mjs
@@ -142,12 +142,6 @@ function safeStringify(arg) {
         );
       }
 
-      errorLogger.error(
-        {
-          reason: reason instanceof Error ? reason.message : safeStringify(reason),
-        },
-        "unhandledRejection"
-      );
     } catch (_) {}
 
     originalConsole.error("Unhandled Promise Rejection:", reason);

